### ISSUE TITLE
Add new module: clustering

### ISSUE THREAD

**Issue by idomic** [ploomber/sklearn-evaluation#52]
We need to create a module for regression plots (can leverage the existing layout from [classification](https://github.com/ploomber/sklearn-evaluation/blob/master/src/sklearn_evaluation/plot/classification.py))

**Comment by edublancas** [comment]
Updates:

added a contributing.md with some guidelines: https://github.com/ploomber/sklearn-evaluation/blob/master/CONTRIBUTING.md

we can get started with the ones that appear here: https://www.scikit-yb.org/en/latest/oneliners.html#clustering

**Comment by edublancas** [comment]
Update: created the clustering module and added one plot https://github.com/ploomber/sklearn-evaluation/blob/master/src/sklearn_evaluation/plot/clustering.py

**Comment by hcho1111** [comment]
TLDR;

1. Is there any way to address the changes between distance based clusters (visually) based on the hyper parameters passed in? Is this worth addressing? 

2. When data is multi-dimensional, clustering doesn’t effectively visualize data. Does it make sense to incorporate some form of dimensionality reduction (maybe PCA) to reduce the complexity of our model? To supplement, we could also include some info on the features in our principal components to give a better understanding on the determinants of our cluster. 

3. I think it may be best to consider two separate modules for clustering - distance based clustering and density-based clustering. 

———

Throughout last week, I took some time to look into clustering methodologies (to become more familiar with methods outside of K-Means). There are a few things that I would like your opinions on such that our viz tools can be effective. 

The first question I have revolves around comparing changes between clusters. To sum, the results of clustering tend to vary significantly. There are a few factors that tend to affect the results of clustering: 

	1. The context of the data you are clustering 
	2. The hyperparameters that are passed during training 
	3. The data normalization methods (or lack thereof) that take place before passing the data to the clustering algo

Assuming that points 1 and 3 are satisfied, do you think it is work implementing some sort of function that allows you to compare the results from cluster visuals? There are more popular hyperparameters that can be addressed via validation plots (i.e. you can pick K in K-Means using the elbow method), but for more complex methods like agglomerative clustering and GMM, there are other factors that may be important to visualize. I’m not sure if this is an issue outside the scope of sklearn-evaluation (optimal metrics could be calculated via CV to ‘achieve’ the best cluster from a universe of user-defined hyperparameters), but I figured that I would ask. 

The second question is slightly connected. Some methods for clustering tend to work better depending on the dimensionality of the dataset. For example, Generative Methods like GMM tend to work better for smaller datasets given the assumptions that are used to calculate the means/covariance matrices of each cluster. Would it make sense to incorporate an additional dimensionality reduction method (like PCA)  to visualize clusters from higher dimensional spaces? I think that it would be "more pleasant" to provide as a visual aid. Obviously, you tend to lose information/ structure about the points  when you apply dimensionality reduction methods, but for the sake of interpretation we could also include the features that affect the specific principal components. Just a thought. 

Finally, on the subject of dimensionality, it may be a good idea to develop a distinction between distance-based and density-based clustering methods. Density-based approaches like DBSCAN, HDBSCAN, and OPTICS tend to function better in higher dimensional spaces and with mixed data types (this falls under issue 3  for distance-based methods). There is a caveat, however, Specifically, HDBSCAN is not native to sklearn. Despite this, however, popular packages (hdbscan,  DenseClus) utilize sklearn scaffolding to create this model. Again, the mathematics and justifications for applying these methods are different compared to distance-methods, so I would recommend drawing a distinction if we choose to include them. This is future work, but I think it would be beneficial to mention as sklearn-eval continues to be built out. 

Wall of text, but would love to hear your thoughts.

**Comment by edublancas** [comment]
I think let's start with something and build from there. did you look at the intercluster distance from here? what about implementing that one? https://www.scikit-yb.org/en/latest/oneliners.html#intercluster-distance

by the way, @neelasha23 is working on the silhouette one.

**Comment by idomic** [comment]
@neelasha23 is there anything that's missing on clustering or we can close this?

**Comment by idomic** [comment]
Closing as added all clustering plots
